### ISSUE TITLE
User Interface considerations in Rack vs hardware modules

### ISSUE THREAD

**Issue by cschol** [VCVRack/Rack#1064]
This is about certain user interface considerations in the context of Rack as opposed to a real hardware module. I am looking for some guidance as it relates to those considerations especially when trying to port an existing module.

For example: 
- pressing and holding a button, then pressing another button
- pressing and holding a button, then moving a knob

I did prototype a solution for both of these scenarios that involved using the `SHIFT` key to "hold" the button. I press `SHIFT`, then click the first button to hold it, then press the next button or move the knob. The module has logic to keep track of the state of the `SHIFT` key and the overall button state per button (i.e. held or just clicked). It actually works pretty well and allows to use the module (4ms Dual Looping Delay) pretty much like the real one.

Is that an acceptable solution?

**Comment by AndrewBelt** [comment]
I personally usually just put all these functions in context menus.

**Comment by cschol** [comment]
In the DLD case, there is some sort of a windowing function where one boundary is set by the knob and the other boundary by the knob while a button is pressed. From a "performance" point of view, when using the module, the `SHIFT` solution feels way more natural than changing the knob mode in the context menu. I tried that first.

Another function is to hold two buttons simultaneously to clear the loop memory. There, the context menu solution is no problem for the workflow and works well.

Here is another problem I ran into with the `SHIFT` solution. I couldn't figure out how to get ahold of the `SHIFT` key other than setting a new handler in my module with `glfwSetKeyCallback` (on the global window instance). That, however, overrides Rack's key handler and none of the Rack keys work anymore. Is there a better way to get a callback when the `SHIFT` key is pressed that does not interfere with other Rack keys?

**Comment by Chaircrusher** [comment]
Those kind of user interface decisions in the hardware domain come about because adding every button or knob is an incremental cost added to the module.

You are not limited to what the hardware you're emulating does and in fact I see it as counterproductive to even try.  Just add a new control that's stateless -- i.e. doesn't depend on another button being pushed.

There is such a thing as going to far down the skeumorphism road. What you describe echoes the hardware, but in fact most of the users of your module won't ever have used the original hardware; there's none of the muscle memory so important to working with modular synthesizer.  What they need is clear, unambiguous controls.

"Here is another problem I ran into with the SHIFT solution" -- I don't know exactly what the situation is internally with key events in VCV Rack, but it sounds like you're swimming upstream against a VERY strong current.

**Comment by cschol** [comment]
@Chaircrusher I agree to a certain extent. Part cost is not the only reason for UI decisions. HP is another and even in the virtual world, HP matters to a certain extent. Not as much as in a physical module, but if I don't want to scroll around in my Rack I have similar limitations as in the physical system.
Like I said above, there are cases where going away from the original design makes sense. In other cases, I would disagree.
